### ISSUE TITLE
no-dupe-keys - warn about non-static key names with similar AST

### ISSUE THREAD

**Issue by victor-homyakov** [eslint/eslint#10724]
**What rule do you want to change?**
`no-dupe-keys`

**Does this change cause the rule to produce more or fewer warnings?**
More

**How will the change be implemented? (New option, new default behavior, etc.)?**
Right now rule works only with static property names https://github.com/eslint/eslint/blob/master/lib/rules/no-dupe-keys.js#L122

Rule could also check for calculated property names with similar AST

**Please provide some example code that this change will affect:**

```js
/* eslint-disable no-console */
const obj1 = {
    ['aaa']: 1,
    ['aaa']: 1 // here duplicate keys are detected
};

const name = 'aaa';

const obj2 = {
    [name]: 1,
    [name]: 1 // here duplicate keys are not detected
};

console.log(obj1, obj2);
```

**What does the rule currently do for this code?**
Rule detects duplicate keys only in `obj1`
```
  4:6  error  Duplicate key 'aaa'  no-dupe-keys

✖ 1 problem (1 error, 0 warnings)
```

**What will the rule do after it's changed?**
The rule will detect also duplicate keys in `obj2`
```
  4:6  error  Duplicate key 'aaa'  no-dupe-keys
 11:6  error  Duplicate key [name]  no-dupe-keys

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by Aladdin-ADD** [comment]
thanks for the issue! 

seems reasonable to me. but I'd prefer the error message was something like `Duplicate key [name]`, for easier to implement and understand.

**Comment by platinumazure** [comment]
Hi @victor-homyakov, thanks for the issue!

At this point, the rule is currently working as designed, so this is an enhancement request.

It would really help us out if you could fill out our [rule enhancement proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md) and lay out a clear change proposal. Could you please edit the initial post with that template? Thanks!

**Comment by platinumazure** [comment]
Thanks @victor-homyakov!

This idea seems like a reasonable enhancement in general, but it would need to be done as an option- otherwise it's a breaking change and we would need to do it in a semver-major release.

Also, we would need to figure out the best way to do this (i.e., what constitutes a "similar AST").

I'm :+1: if it's done as an option.

**Comment by ljharb** [comment]
Is there a reason this wasn't accepted? Can it be revisited?
